partition-manager/generation/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+        pub variants: BTreeSet<Variant>,
+        pub disk: Disk,
+        pub partitions: BTreeMap<String, Partition>,
+    }
+
+    impl Manifest {
+        fn validate_alignment(&self) -> anyhow::Result<()> {
+            if let Some(alignment) = self.disk.alignment {
+                for (name, partition) in self.partitions.iter() {
+                    if partition.offset % alignment != 0 || partition.size % alignment != 0 {
+                        return Err(anyhow!("Partition {} is not aligned to {} bytes", name, alignment));
+                    }
+                }
+            }
+            Ok(())
+        }
+
+        fn validate_overlap(&self) -> anyhow::Result<()> {
+            let mut partitions = Vec::from_iter(self.partitions.iter());
+            partitions.sort_by_key(|(_, partition)| partition.offset);
+
+            for (i, (partition_name_x, partition_x)) in partitions.iter().enumerate() {
+                for (partition_name_y, partition_y) in partitions[(i + 1).min(partitions.len())..].iter() {
+                    if partition_x.overlaps(partition_y) {
+                        return Err(anyhow!(
+                            "Partitions {} and {} overlap",
+                            partition_name_x,
+                            partition_name_y
+                        ));
+                    }
+                }
+            }
+
+            Ok(())
+        }
+
+        fn validate_size(&self) -> anyhow::Result<()> {
+            if let Some(size) = self.disk.size {
+                for (name, partition) in self.partitions.iter() {
+                    if partition.offset + partition.size > size {
+                        return Err(anyhow!("Partition {} goes over underlying disk edge", name));
+                    }
+                }
+            }
+            Ok(())
+        }
+
+        pub fn check_consistency(&self) -> anyhow::Result<()> {
+            self.validate_size()?;
+            self.validate_overlap()?;
+            self.validate_alignment()?;
+            Ok(())
+        }
+
+        pub fn generate(
+            self,
+            variant_name: Option<String>,
+        ) -> anyhow::Result<impl Iterator<Item = GeneratedPartition>> {
+            self.check_consistency()?;
+
+            let variant_name = match variant_name {
+                Some(variant_name) => {
+                    let variant_name = variant_name.to_string();
+                    if !self.variants.contains(&Variant::Other(variant_name.clone())) {
+                        return Err(anyhow!("Variant '{}' not defined in manifest", variant_name));
+                    }
+                    Some(variant_name)
+                }
+                None => None,
+            };
+
+            let variant = variant_name.map(Variant::Other).unwrap_or(Variant::Any);
+
+            Ok(self
+                .partitions
+                .into_iter()
+                .filter_map(move |(partition_name, partition)| {
```

Comment:
Nit-Pick: You could destructure `partition` here, making copying `access`, `name`, and `size` a tiny-bit less verbose.

But feel free to leave it as is.

Reply:
Applied!